fiat-roles/src/main/java/com/netflix/spinnaker/fiat/providers/internal/resourcegroups/BaseGroupResolutionStrategy.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2019 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.fiat.providers.internal.resourcegroups;
+
+import com.netflix.spinnaker.fiat.model.resources.Permissions;
+import com.netflix.spinnaker.fiat.model.resources.Resource;
+import com.netflix.spinnaker.fiat.model.resources.groups.ResourceGroup;
+import java.util.Set;
+import java.util.stream.Collectors;
+
```

Comment:
@cfieber I am not very happy with what I'm doing here.
I don't like that we give the resolution strategy the whole, unfiltered groups, and it takes care of the filtering. I would rather give it pre-filtered groups, and just get rid of that `BaseResolutionStrategy` class, but I don't know where to put the filtering logic. Do you have any suggestions?

Reply:
I was thinking a bit about this, and with the move away from front50 storing the ResourceGroups I don't think we need to keep a single jackson-deserializable hierarchy anymore.

I think we could make a ResourceGroupProvider (similar to ResourceProvider), make those Spring components, and they could independently get configuration (via `@ConfigurationProperties`) for the rules they need.

This is maybe similar to where I was headed in https://github.com/spinnaker/fiat/pull/472 except s/ResourceInterceptor/ResourceGroupProvider/ and also supporting the lookup of permissions for a resource that does not yet exist.

We can then autowire in a collection of all ResourceGroupProviders that exist, and each of them can augment permissions for whatever resources they support. This also has the added benefit of being able to extend/add new ResourceGroupProviders via plugins (right now the ResourceGroupType / jackson deserialization makes that kind of impossible).

Reply:
Hmmmm I think I've understood most of what you said except the last paragraph.
Having a `ResourceGroupProvider` is indeed a good idea. It is already kind of implemented right now in `FiatResourceGroupConfig`, but I can move the logic out to a separate provider and make things more generic. Let me see what I can do